demo/lib/topology.py

```diff
@@ -0,0 +1,272 @@
+                    new_line.append(txt)
+                    new_col_txt_added = True
+                else:
+                    new_line.append("")
+            lines.append(new_line)
+        branch.pop()
+
+def str_tree(root):
+    """Format tree to string."""
+    lines = []
+    _str_node(root, lines, [])
+    col_max_len = {} # {column-index: max-string-length}
+    max_col = -1
+    for line in lines:
+        for col, txt in enumerate(line):
+            if col > max_col:
+                max_col = col
+            if len(txt) > col_max_len.get(col, -1):
+                col_max_len[col] = len(txt)
+    str_lines = []
+    for line in lines:
+        line_cols = len(line)
+        new_str_fmt = ""
+        for col, txt in enumerate(line):
+            new_str_fmt += "%-" + str(col_max_len[col] + 1) + "s"
+        str_lines.append(new_str_fmt % tuple(line))
+    return "\n".join(str_lines)
+
+def bash_output(cmd):
+    """Return standard output of executing cmd in Bash."""
+    p = subprocess.Popen(["bash", "-c", cmd], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    out, err = p.communicate()
+    return out.decode("utf-8")
+
+def get_local_topology_dump():
+    """Return topology_dump from local system."""
+    return bash_output(_bash_topology_dump)
+
+def get_local_cpus_allowed_dump(processes):
+    """Return cpus_allowed from local system."""
+    return bash_output(_bash_cpus_allowed % ("' '".join(processes),))
+
+def dump_to_topology(dump):
+    """Parse topology_dump, return topology data structures."""
+    # Output data structures:
+    tree = {} # {"package0": {"die1": {"node1": ...}}}
+    cpu_branch = {} # {cpu_id: (package_name, die_name, node_name, core_name, thread_name, cpu_name)}
+    # Example input line to be parsed:
+    # "cpu p:0 d:1 n:3 c:2 t:00003000 cpu:13"
+    re_dump_line = re.compile('cpu p:(?P<package>[0-9]+) d:(?P<die>[0-9]+) n:(?P<node>[0-9]+) c:(?P<core>[0-9]+) t:(?P<thread_siblings>[0-9a-f,]+) cpu:(?P<cpu_id>[0-9]+)')
+    for line in dump.splitlines():
+        try:
+            mdict = re_dump_line.match(line).groupdict()
+        except:
+            continue
+        package = int(mdict["package"])
+        die = int(mdict["die"])
+        node = int(mdict["node"])
+        core = int(mdict["core"])
+        thread_siblings = eval("0x" + mdict["thread_siblings"].replace(",", ""))
+        cpu_id = int(mdict["cpu_id"])
+        # Calculate thread id.
+        # Let the lowest CPU bit owner in thread_siblings be thread 0, next thread 1 and so on.
+        thread = -1
+        bit = 1 << cpu_id
+        while bit:
+            if thread_siblings & bit:
+                thread += 1
+            bit >>= 1
+        branch = ("package" + str(package),
+                  "die" + str(die),
+                  "node" + str(node),
+                  "core" + str(core).zfill(3),
+                  "thread" + str(thread),
+                  "cpu" + str(cpu_id).zfill(3))
```

Comment:
nitpick: since everything else is carefully aligned and repetitive duplicate printouts filtered, why not also to zfill only to len(str(max_core_id)) and len(str(max_cpu_id)) ?

Reply:
Good point. Now all numbers are numbers are formatted similarly: zfilled to minimum necessary length. And I used `len(str(n))` over `int(math.log(10))+1` for readability. (Resisted 2x perf difference... these are tiny data structures anyway.)